jakarta-eclipse/src/org/jakarta/jdt/cdi/ManagedBeanDiagnosticsCollector.java

```diff
@@ -0,0 +1,99 @@
+    private static final String DIAGNOSTIC_SOURCE = "jakarta-cdi";
+    private static final DiagnosticSeverity SEVERITY = DiagnosticSeverity.Error;
+    private static final Set<String> SCOPES = new HashSet<String>(
+            Arrays.asList("Dependent", "ApplicationScoped", "ConversationScoped", "RequestScoped", "SessionScoped"));
+
+    private Diagnostic createDiagnostic(ICompilationUnit unit, IJavaElement el, String message)
+            throws JavaModelException {
+        ISourceRange nameRange = JDTUtils.getNameRange(el);
+        Range range = JDTUtils.toRange(unit, nameRange.getOffset(), nameRange.getLength());
+        Diagnostic diagnostic = new Diagnostic(range, message);
+        completeDiagnostic(diagnostic);
+        return diagnostic;
+    }
+
+    @Override
+    public void completeDiagnostic(Diagnostic diagnostic) {
+        diagnostic.setSource(DIAGNOSTIC_SOURCE);
+        diagnostic.setSeverity(SEVERITY);
+    }
+
+    @Override
+    public void collectDiagnostics(ICompilationUnit unit, List<Diagnostic> diagnostics) {
+        if (unit == null)
+            return;
+
+        List<String> managedBeanAnnotations;
+
+        try {
+            for (IType type : unit.getAllTypes()) {
+                // Construct a stream of only the annotations applied to the type that are also
+                // recognised managed bean annotations.
+                managedBeanAnnotations = Arrays.stream(type.getAnnotations())
+                        .map(annotation -> annotation.getElementName()).filter(SCOPES::contains).distinct()
+                        .collect(Collectors.toList());
+
+                boolean isManagedBean = managedBeanAnnotations.size() > 0;
+
+                for (IField field : type.getFields()) {
+                    int fieldFlags = field.getFlags();
+
+                    /**
+                     * If a managed bean has a non-static public field, it must have
+                     * scope @Dependent. If a managed bean with a non-static public field declares
+                     * any scope other than @Dependent, the container automatically detects the
+                     * problem and treats it as a definition error.
+                     * 
+                     * https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html#managed_beans
+                     */
+                    if (isManagedBean && Flags.isPublic(fieldFlags) && !Flags.isStatic(fieldFlags)
+                            && managedBeanAnnotations.stream()
+                                    .anyMatch(annotation -> !annotation.equals("Dependent"))) {
+                        Diagnostic diagnostic = createDiagnostic(unit, field,
+                                "If a managed bean has a non-static public field, it must have scope @Dependent.\n"
+                                        + "A managed bean with a non-static public field must also not declare any scope other than @Dependent");
```

Comment:
These two lines seem to be saying the same thing. Consider just the second statement: "A managed bean with a non-static public field must also not declare any scope other than `@Dependent`"

Reply:
Re-reading that statement, let's remove the "also" so that it reads:
"A managed bean with a non-static public field must not declare any scope other than `@Dependent`"